src/directory/s3_directory.rs

```diff
@@ -0,0 +1,248 @@
+use atomicwrites;
+use common::make_io_err;
+use directory::Directory;
+use directory::error::{IOError, OpenWriteError, OpenReadError, DeleteError, OpenDirectoryError};
+use directory::ReadOnlySource;
+use directory::shared_vec_slice::SharedVecSlice;
+use directory::WritePtr;
+use fst::raw::MmapReadOnly;
+use memmap::{Mmap, Protection};
+use std::collections::hash_map::Entry as HashMapEntry;
+use std::collections::HashMap;
+use std::convert::From;
+use std::default::Default;
+use std::fmt;
+use std::fs::{self, File};
+use std::fs::OpenOptions;
+use std::io::{self, Seek, SeekFrom};
+use std::io::{BufWriter, Read, Write};
+use std::mem;
+use std::path::{Path, PathBuf};
+use std::result;
+use std::str::FromStr;
+use std::sync::Arc;
+use std::sync::RwLock;
+use std::sync::Weak;
+use tempdir::TempDir;
+use rusoto_core::{DefaultCredentialsProvider, Region, default_tls_client};
+use rusoto_s3::{S3, S3Client, HeadBucketRequest, GetObjectRequest};
+
+#[derive(Clone)]
+struct InnerDirectory(Arc<RwLock<HashMap<PathBuf, Arc<Vec<u8>>>>>);
+
+impl InnerDirectory {
+    fn new() -> InnerDirectory {
+        InnerDirectory(Arc::new(RwLock::new(HashMap::new())))
+    }
+}
+
+/// Directory storing data in files, read via mmap.
+///
+/// The Mmap object are cached to limit the
+/// system calls.
+#[derive(Clone)]
+pub struct S3Directory {
```

Comment:
I think I want to cache the `S3` client here rather than just create it all the time.